packages/ckeditor5-ui/src/dropdown/utils.ts

```diff
@@ -541,6 +543,24 @@ function bindViewCollectionItemsToDefinitions(
 	definitions: Collection<ListDropdownItemDefinition>,
 	locale: Locale
 ) {
+	// List item checkboxes have a reserved space for the check icon, so we need to know if there are any checkboxes in the list
+	// to adjust the layout accordingly. It'd look weird if the items on the list were not aligned horizontally.
+	listItems.on( 'change', () => {
```

Comment:
Also, I wonder if we could instead listen to `add` and `remove` (or was it `delete`?) events on these collections and maybe keep internal counter of "toggleable" items. Or even simpler, just check if there was ever a toggleable button added, maybe such simplification will be good enough (to be honest - I am quite hesitant whether we need to have it dynamic; I think that it would be enough to set it "on initialization").

Reply:
> For performance reasons, I think we could add this event after the list is populated and also of course evaluate it once for initial load. I am a bit afraid of the longer lists, I am afraid it will have square complexity and may impact performance when a few hundred entries are loaded.

The performance should not be an issue, at least not in this particular listener. I checked it before implementation and `change` is fired [once](https://github.com/ckeditor/ckeditor5/blob/4f4032a74a8fe1ca0e4d6ace27bd660b095c674b/packages/ckeditor5-utils/src/collection.ts#L228-L234), event after calling `addMany`. So it's perfectly fine to assume that for the majority part of UI it'll be called only once, and it'll be fired occasionally while extending an existing dropdown, which is not super common across our UI. On the other hand - if there is and dropdown, that uses `add()` instead of `addMany()`, it can spam a bit and increase complexity. I don't think it matters.

Reply:
If it uses `addMany()` then it is fine. I was worried the items are usually (always) added one by one.

Reply:
Can we add an inline comment:

```
// Possible theoretical performance problem if many items are added one by one, as this will be called for each item.
```

?

Reply:
Comment added